ui/client/components/graph/node-modal/editors/expression/EditableExpression.tsx

```diff
@@ -0,0 +1,85 @@
+import React from "react"
+import ProcessUtils from "../../../../../common/ProcessUtils"
+import {dualEditorMode, editors, editorTypes} from "./EditorType"
+import SwitchIcon from "./SwitchIcon"
+import FixedValuesEditor from "./FixedValuesEditor"
+import _ from "lodash";
+
+type Props = {
+  fieldType: string
+  expressionObj: any
+  showSwitch: boolean
+  renderFieldLabel: Function
+  fieldLabel: string
+  readOnly: boolean
+  rowClassName?: string
+  valueClassName?: string
+  param?: any
+  values?: Array<any>
+  editorName?: string
+  fieldName?: string
+  isMarked?: boolean
+  validators?: Array<any>
+  showValidation?: boolean
+  onValueChange: Function
+}
+
+type State = {
+  displayRawEditor: boolean
+}
+
+class EditableExpression extends React.Component<Props, State> {
+
+  state = {
+    displayRawEditor: undefined
+  }
+
+  toggleEditor = (_) => {
+    const {param} = this.props
+    this.setState({
+      displayRawEditor: this.state.displayRawEditor === undefined ?
+        (param.editor.defaultMode === dualEditorMode.SIMPLE) :
+        !this.state.displayRawEditor
+    })
+  }
+
+  render() {
+    const {
+      fieldType, expressionObj, rowClassName, valueClassName, showSwitch, param, renderFieldLabel, fieldLabel, readOnly,
+      values
+    } = this.props
+
+    const paramType = fieldType || (param ? ProcessUtils.humanReadableType(param.typ.refClazzName) : "expression")
+
+    const editorType = paramType === editorTypes.FIXED_VALUES_PARAMETER_EDITOR ?
+      editorTypes.FIXED_VALUES_PARAMETER_EDITOR :
+      (!_.isEmpty(param) ? param.editor.type : editorTypes.RAW_PARAMETER_EDITOR)
+    const editor = editors[editorType]
+
+    const Editor = editor.editor(param, this.state.displayRawEditor)
+
+    return (
+      <div className={`${rowClassName ? rowClassName : " node-row"}`}>
+        {fieldLabel && renderFieldLabel(fieldLabel)}
+        <Editor toggleEditor={this.toggleEditor}
+                className={`${valueClassName ? valueClassName : "node-value"} ${editor.showSwitch ? "switchable " : ""}`}
+                {...this.props}
+                values={Editor === FixedValuesEditor ? editor.values(param, values) : []}
+        />
+        {
+          _.get(param, "editor.type") === editorTypes.DUAL_PARAMETER_EDITOR &&
```

Comment:
lodash not needed, we prefer optional chaining in ts

Reply:
Done